core-bundle/src/InsertTag/InsertTagParser.php

```diff
@@ -434,16 +493,84 @@ private function renderSubscription(InsertTag $tag, bool $allowEsiTags): InsertT
         $result = $subscription->service->{$subscription->method}($tag);
 
         foreach ($tag->getFlags() as $flag) {
+            // ESI tags need to be replaced before flags can be applied
+            $result = $this->replaceEsiTags($result, $esiTagsCount);
+
+            if ($esiTagsCount) {
+                $this->logger->error(
+                    \sprintf(
+                        'Using the insert tag flag "%s" in %s on page %s disables lazy loading of the fragment',
+                        $flag->getName(),
+                        $tag->serialize(),
+                        $this->framework->getAdapter(Environment::class)->get('uri'),
+                    ),
+                );
+            }
+
             if ($callback = $this->flagCallbacks[strtolower($flag->getName())] ?? null) {
                 $result = $callback($flag, $result);
             } else {
                 $result = $this->handleLegacyFlagsHook($result, $flag, $tag);
             }
         }
 
+        if (!$allowEsiTags) {
+            $result = $this->replaceEsiTags($result);
+        }
+
         return $result;
     }
 
+    /**
+     * @see \Symfony\Component\HttpKernel\HttpCache\Esi::process()
+     *
+     * @param-out int $esiTagsCount
+     */
+    private function replaceEsiTags(InsertTagResult $result, int|null &$esiTagsCount = 0): InsertTagResult
```

Comment:
This would also happen automatically within the kernel if you send it through the entire stack using a sub request?